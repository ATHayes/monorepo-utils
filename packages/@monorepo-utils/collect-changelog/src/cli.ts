// MIT Â© 2018 azu
"use strict";

import * as fs from "fs";

import * as path from "upath";
import { findChangelog, getChangelog } from "./collect-changelog-from-tag";
import Handlebars = require("handlebars");

export interface ExecuteOptions {
    directory: string;
    changelogFilePath?: string;
    name?: string;
    tag: string;
    templatePath?: string;
}

export async function execute(options: ExecuteOptions) {
    const templatePath = options.templatePath
        ? path.resolve(process.cwd(), options.templatePath)
        : path.join(__dirname, "../template/changelog.handlebars");
    const code = fs.readFileSync(templatePath, "utf-8");
    const template = Handlebars.compile(code);
    if (options.changelogFilePath) {
        const changelog = await getChangelog(options.changelogFilePath, {
            name: options.name ?? "package", // no name
            version: options.tag,
        });
        return template(changelog).trim();
    } else {
        const changelog = await findChangelog(options.directory, options.tag);
        return template(changelog).trim();
    }
}
