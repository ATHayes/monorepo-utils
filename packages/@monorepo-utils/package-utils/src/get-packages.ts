"use strict";

const globby = require("globby");
import loadJsonFile = require("load-json-file");
import * as path from "path";
import * as fs from "fs";

const loadPackage = (packagePath: string) => {
    const pkgJsonPath = path.join(packagePath, "package.json");
    if (fs.existsSync(pkgJsonPath)) {
        return loadJsonFile.sync(pkgJsonPath);
    }
};

const findPackages = (packageSpecs: string[], rootDirectory: string) => {
    return packageSpecs
        .reduce(
            (pkgDirs, pkgGlob) => {
                const target = path.join(rootDirectory, pkgGlob);
                const filePathList = globby.hasMagic(pkgGlob)
                    ? globby.sync(target, {
                          onlyFiles: false
                      })
                    : [target];
                return pkgDirs.concat(filePathList);
            },
            [] as string[]
        )
        .map((location: string) => {
            return { location, package: loadPackage(location) };
        })
        .filter(res => {
            return res.package && res.package.name;
        });
};

export const getPackages = (directory: string) => {
    const lernaJsonPath = path.join(directory, "lerna.json");
    if (fs.existsSync(lernaJsonPath)) {
        const lernaJson = loadJsonFile.sync(lernaJsonPath);
        if (!lernaJson.useWorkspaces) {
            return findPackages(lernaJson.packages, directory);
        }
    }

    const pkgJsonPath = path.join(directory, "package.json");
    if (fs.existsSync(pkgJsonPath)) {
        const pkgJson = loadJsonFile.sync(pkgJsonPath);
        if (pkgJson.workspaces) {
            // "workspaces": []
            if (Array.isArray(pkgJson.workspaces)) {
                return findPackages(pkgJson.workspaces, directory);
            } else if (Array.isArray(pkgJson.workspaces.packages)) {
                // "workspaces": { "packages": [] }
                return findPackages(pkgJson.workspaces.packages, directory);
            }
        }
    }

    // Bail if we don't find any packages
    return [];
};
