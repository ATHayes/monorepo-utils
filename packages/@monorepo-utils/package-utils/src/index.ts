export { getPackages } from "./get-packages";
export type { PackageResult } from "./get-packages";
