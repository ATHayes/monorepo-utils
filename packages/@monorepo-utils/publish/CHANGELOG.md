# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="1.3.0"></a>
# [1.3.0](https://github.com/azu/monorepo-utils/compare/@monorepo-utils/publish@1.2.2...@monorepo-utils/publish@1.3.0) (2018-08-06)


### Features

* **publish:** support --dist-tag ([d00662b](https://github.com/azu/monorepo-utils/commit/d00662b))




<a name="1.2.2"></a>
## [1.2.2](https://github.com/azu/monorepo-utils/compare/@monorepo-utils/publish@1.2.1...@monorepo-utils/publish@1.2.2) (2018-08-03)


### Bug Fixes

* **publish:** fix bin command ([27baef7](https://github.com/azu/monorepo-utils/commit/27baef7))




<a name="1.2.1"></a>
## [1.2.1](https://github.com/azu/monorepo-utils/compare/@monorepo-utils/publish@1.2.0...@monorepo-utils/publish@1.2.1) (2018-08-03)




**Note:** Version bump only for package @monorepo-utils/publish

<a name="1.2.0"></a>
# [1.2.0](https://github.com/azu/monorepo-utils/compare/@monorepo-utils/publish@1.1.0...@monorepo-utils/publish@1.2.0) (2018-07-22)


### Features

* **publish:** add version to publish log ([6c9d70a](https://github.com/azu/monorepo-utils/commit/6c9d70a))




<a name="1.1.0"></a>
# 1.1.0 (2018-07-22)


### Bug Fixes

* **publish:** fix usage ([047968a](https://github.com/azu/monorepo-utils/commit/047968a))


### Features

* **package-utils:** add package-utils ([3c49df9](https://github.com/azu/monorepo-utils/commit/3c49df9))
