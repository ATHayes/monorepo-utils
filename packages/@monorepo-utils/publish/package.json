{
  "name": "@monorepo-utils/publish",
  "version": "1.2.0",
  "description": "Publish packages to npm if needed. ",
  "license": "MIT",
  "files": [
    "publish.js",
    "utils"
  ],
  "main": "lib/publish.js",
  "types": "lib/publish.d.ts",
  "bin": {
    "monorepo-utils-publish": "publish.js"
  },
  "scripts": {
    "build": "cross-env NODE_ENV=production tsc -p .",
    "prepublish": "npm run --if-present build",
    "test": "jest",
    "watch": "tsc -p . --watch"
  },
  "dependencies": {
    "@monorepo-utils/package-utils": "^1.1.0",
    "can-npm-publish": "^1.3.1",
    "chalk": "^2.3.0",
    "child-process-promise": "^2.2.1",
    "cli-spinners": "^1.1.0",
    "inquirer": "^6.0.0",
    "log-update": "^2.3.0",
    "meow": "^5.0.0"
  },
  "devDependencies": {
    "@types/jest": "^23.3.1",
    "@types/node": "^10.5.5",
    "consolemock": "^1.1.0",
    "cross-env": "^5.2.0",
    "jest": "^23.4.2",
    "ts-jest": "^23.1.0",
    "typescript": "^3.0.1"
  },
  "publishConfig": {
    "access": "public"
  },
  "jest": {
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "jsx",
      "json",
      "node"
    ],
    "testRegex": "./test/.*.ts$",
    "rootDir": ".",
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    },
    "globals": {
      "ts-jest": {
        "skipBabel": true
      }
    }
  }
}
