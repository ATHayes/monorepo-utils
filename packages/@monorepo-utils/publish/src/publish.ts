const inquirer = require("inquirer");
const { getPackages } = require("@monorepo-utils/package-utils");
const canNpmPublish = require("can-npm-publish").canNpmPublish;
import { execUnlessDry, log, logPromise } from "./utils/utils";
import { asyncFilter } from "./utils/async-utils";

/**
 * publish packages to npm
 */
export const publish = async ({
    projectDir,
    dry,
    skipPrompt,
    ciMode
}: {
    projectDir: string;
    dry: boolean;
    skipPrompt: boolean;
    ciMode: boolean;
}): Promise<void> => {
    const packages = getPackages(projectDir);
    const publishablePackages = await asyncFilter(packages, async (pkg: any) => {
        try {
            await canNpmPublish(pkg.location);
            return true;
        } catch (error) {
            return false;
        }
    });
    if (publishablePackages.length === 0) {
        return log("No need to publish.");
    }
    // confirm: npm publish if skipPrompt is false
    const answers = skipPrompt
        ? {
              publish: true
          }
        : await inquirer.prompt([
              {
                  type: "confirm",
                  name: "publish",
                  message: `Do you publish these packages?
${publishablePackages
                      .map((pkg: any) => {
                          return `- ${pkg.package.name}@${pkg.package.version}`;
                      })
                      .join("\n")}
`
              }
          ]);
    if (!answers["publish"]) {
        return log("No publish");
    }
    // npm publish
    await publishablePackages.reduce((promise: Promise<any>, pkg: any) => {
        return promise.then(() => {
            const location = pkg.location;
            const task = execUnlessDry(`npm publish`, {
                cwd: location,
                dry
            });
            return logPromise(task, `${pkg.package.name}@${pkg.package.version}`, {
                ciMode: ciMode
            });
        });
    }, Promise.resolve());
};
