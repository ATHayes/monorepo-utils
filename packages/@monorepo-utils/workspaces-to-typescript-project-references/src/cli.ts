import meow from "meow";
import path from "path";
import { toProjectReferences } from "./index";

export const cli = meow(
    `
    Usage
      $ workspaces-to-typescript-project-references
 

    Options
      --root             [Path:string] Root directory of the monorepo. 
                         Default: current working directory
      --check            If set the flag, check only differences of tsconfig.json and does not update tsconfig.json.
                         If the check is failed, exit status 1. It is useful for testing.
       
      --plugin           [Path:string] Path to plugin script.
                         Load the plugin script as module and use it. 
                           
    Examples
      # Update project references in tsconfig.json
      $ workspaces-to-typescript-project-references
      # Test on CI
      $ workspaces-to-typescript-project-references --check
`,
    {
        flags: {
            root: {
                type: "string",
                default: process.cwd()
            },
            check: {
                type: "boolean",
                default: false
            },
            plugin: {
                type: "string",
                isMultiple: true
            }
        },
        autoHelp: true,
        autoVersion: true
    }
);

export const run = async (
    _input = cli.input,
    flags = cli.flags
): Promise<{ exitStatus: number; stdout: string | null; stderr: string | null }> => {
    const plugins = Array.isArray(flags.plugin)
        ? flags.plugin.map((pluginPath) => require(path.join(process.cwd(), pluginPath)))
        : undefined;
    const result = toProjectReferences({
        rootDir: flags.root,
        checkOnly: flags.check,
        plugins
    });
    if (result.ok) {
        return {
            exitStatus: 0,
            stdout: flags.check ? "" : "Update Project References!",
            stderr: null
        };
    } else if (result.aggregateError) {
        return {
            exitStatus: 1,
            stdout: null,
            stderr: result.aggregateError.message + "\n\n" + result.aggregateError.errors.join("\n")
        };
    }
    return {
        exitStatus: 0,
        stdout: null,
        stderr: null
    };
};
