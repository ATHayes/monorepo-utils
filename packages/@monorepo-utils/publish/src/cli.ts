#!/usr/bin/env node

import { publish } from "./publish";
import { execRead, getUnexecutedCommands } from "./utils/utils";

const chalk = require("chalk");
const meow = require("meow");

const cli = meow(
    `
    Usage
      $ monorepo-utils-publish

    Options:
      --dry Enable dry run mode
      --yes Yes all
      --ci  CI mode includes --yes flag

    Examples
      # dry-run mode
      $ monorepo-utils-publish --dry
      # manually publish
      # CI mode
      $ monorepo-utils-publish --ci
`,
    {
        autoVersion: true,
        autoHelp: true,
        flags: {
            dry: {
                type: "boolean"
            },
            yes: {
                type: "boolean"
            },
            ci: {
                type: "boolean"
            }
        }
    }
);

const dryRunMode = !!cli.flags.dry;
if (dryRunMode) {
    console.log(
        chalk`
    {green.bold dry-run mode}
    `
    );
}

export async function run() {
    const projectDir = await execRead("git rev-parse --show-toplevel");
    const ciMode = !!cli.flags.ci;
    await publish({
        projectDir,
        dry: dryRunMode,
        skipPrompt: ciMode ? true : !!cli.flags.yes,
        ciMode: ciMode
    });
    console.log(
        chalk`
    {green.bold Publish successful!}
    ${getUnexecutedCommands()}
  `.replace(/\n +/g, "\n")
    );
}
