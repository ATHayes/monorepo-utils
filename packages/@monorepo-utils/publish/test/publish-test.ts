import * as path from "path";
import { publish } from "../src/publish";

const stripAnsi = require("strip-ansi");
const makeConsoleMock = require("consolemock");
describe("publish", () => {
    let consoleMock: any;
    let originalConsole: any;
    beforeEach(() => {
        consoleMock = makeConsoleMock(console);
        originalConsole = global.console;
        global.console = consoleMock;
    });
    afterEach(() => {
        global.console = originalConsole;
    });
    it("should publish packages", async () => {
        await publish({
            projectDir: path.join(__dirname, "fixtures"),
            skipPrompt: true,
            dry: true,
            ciMode: true
        });
        expect(stripAnsi(consoleMock.printHistory())).toMatchSnapshot();
    });
    it("should publish packages with --distTag", async () => {
        await publish({
            projectDir: path.join(__dirname, "fixtures"),
            skipPrompt: true,
            dry: true,
            ciMode: true,
            distTag: "next"
        });
        expect(stripAnsi(consoleMock.printHistory())).toMatchSnapshot();
    });
});
