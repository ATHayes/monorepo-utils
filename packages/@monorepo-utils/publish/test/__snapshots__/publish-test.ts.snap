// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`publish should publish packages 1`] = `
"
LOG ✓ textlint@1000.0.0-test
LOG ✓ @textlint/ast-node-types@1000.0.0-test
LOG ✓ @textlint/kernel@1000.0.0-test
LOG ✓ @textlint/linter-formatter@1000.0.0-test
LOG ✓ @textlint/markdown-to-ast@1000.0.0-test"
`;
